### DISCUSSION TITLE
Modular proposal in the contracts

### DISCUSSION THREAD

**Discussion by luislucena16** [PaulRBerg/prb-proxy#187] [Ideas]
This is more of a proposal. It's a matter of preference :)... I think you could modularise the [interface](https://github.com/PaulRBerg/prb-proxy/blob/main/src/interfaces/IPRBProxy.sol) and tidy up a bit what's there, maybe errors, events and constants could have a separate contract for each one

PS: in case it grows over time!

**Comment by PaulRBerg** [comment]
Thanks for your suggestion, @luislucena16!

Just to clarify what you mean - do you want something like [this](https://github.com/sablier-labs/v2-core/blob/7819a3b5122036596edff3326b97a8c03750bbb8/src/libraries/Errors.sol) for errors, and do the same for events and constants?

Related: [Use libraries as namespaces for structs and enums](https://twitter.com/PaulRBerg/status/1619312803944480770)

**Reply by luislucena16** [reply]
yes, just like the example of [errors](https://github.com/sablier-labs/v2-core/blob/7819a3b5122036596edff3326b97a8c03750bbb8/src/libraries/Errors.sol)!

**Reply by luislucena16** [reply]
I can also propose that all the `internal` functions and the `struct` could go into a contract that is called `Storage`.

for example: 
- `Proxy1` has `internal` functions and `struct`
- Create `Proxy1Storage` and stores what it is entitled to

**Reply by PaulRBerg** [reply]
Actually, that's how it used to be, but I removed that design in this PR:

https://github.com/PaulRBerg/prb-proxy/pull/120

**Reply by luislucena16** [reply]
I see it! any specific reason of the change?

**Reply by PaulRBerg** [reply]
@luislucena16 yep, see this audit report.

[Plugins and permissions storage variables are unprotected · Issue #13 · cantinasec_review-sablier2.pdf](https://github.com/PaulRBerg/prb-proxy/files/14124287/Plugins.and.permissions.storage.variables.are.unprotected.Issue.13.cantinasec_review-sablier2.pdf)

**Reply by luislucena16** [reply]
okay I see! I understand the point, some things I consider are these:

(i) the change was made because of the vulnerability of them being able to access mapping or variables that somehow end up being stored in the PRBProxyStorage, but I think it could have been handled as `internal` and had something like this:

- this is how you had it:
```bash
abstract contract PRBProxyStorage is IPRBProxyStorage {
address public override owner;
uint256 public override minGasReserve;
mapping(bytes4 method => IPRBProxyPlugin plugin) public plugins;
mapping(address envoy => mapping(address target => bool permission)) public permissions;
```
- so I thought:
```bash
abstract contract PRBProxy is PRBProxyStorage {
address internal override owner;
uint256 internal override minGasReserve;
mapping(bytes4 method => IPRBProxyPlugin plugin) internal plugins;
mapping(address envoy => mapping(address target => bool permission)) internal permissions;
```

the inheritance would be `PRBProxy is PRBProxyStorage` (removing the interface) and what contains `PRBProxyStorage` would be all `internal`.

ii) I think a validation could also be added here to ensure that the address is not a contract to avoid any malicious intent. A function written in assembly that focuses on that validation. What do you think?

```bash
contract PRBProxyRegistry {
function installPlugin(address owner, IPRBProxyPlugin plugin) external override {
if (proxies[owner] != msg.sender) { revert("Unauthorized"); }
// Impossible to check if `msg.sender` is the proxy itself or a malicious delegate-ca
  }
}
```
